openedx/core/djangoapps/content/search/api.py

```diff
@@ -348,13 +348,16 @@ def rebuild_index(status_cb: Callable[[str], None] | None = None) -> None:
             try:
                 docs = []
                 for component in lib_api.get_library_components(lib_key):
-                    metadata = lib_api.LibraryXBlockMetadata.from_component(lib_key, component)
-                    doc = {}
-                    doc.update(searchable_doc_for_library_block(metadata))
-                    doc.update(searchable_doc_tags(metadata.usage_key))
-                    docs.append(doc)
-
-                    num_blocks_done += 1
+                    try:
+                        metadata = lib_api.LibraryXBlockMetadata.from_component(lib_key, component)
+                        doc = {}
+                        doc.update(searchable_doc_for_library_block(metadata))
+                        doc.update(searchable_doc_tags(metadata.usage_key))
+                        docs.append(doc)
+                    except Exception as err:  # pylint: disable=broad-except
+                        status_cb(f"Error indexing library component {component}: {err}")
+                    finally:
```

Comment:
If we're catching expections here, do we need the Exception try/catch that wraps this?

Reply:
The outer try/catch will handle exceptions while calling the meilisearch API to add the documents. They're not likely to fail, but if you don't see a problem, I think it's worth the extra try/catch. What do you think?

Also, @pomegranited, could you CC review this?

Reply:
With a little rearranging, you can avoid the nested try/catch blocks. For the Meilisearch errors, we know what might be thrown, so I'd rather that not be a broad Exception check. Then we've just got the one added here (which makes sense to keep broad).

Like this?

```diff
-# pylint: disable=too-many-statements disable=too-many-nested-blocks
+# pylint: disable=too-many-statements
 def rebuild_index(status_cb: Callable[[str], None] | None = None) -> None:
     """
     Rebuild the Meilisearch index from scratch
@@ -345,26 +345,26 @@ def rebuild_index(status_cb: Callable[[str], None] | None = None) -> None:
         status_cb("Indexing libraries...")
         for lib_key in lib_keys:
             status_cb(f"{num_contexts_done + 1}/{num_contexts}. Now indexing library {lib_key}")
-            try:
-                docs = []
-                for component in lib_api.get_library_components(lib_key):
-                    try:
-                        metadata = lib_api.LibraryXBlockMetadata.from_component(lib_key, component)
-                        doc = {}
-                        doc.update(searchable_doc_for_library_block(metadata))
-                        doc.update(searchable_doc_tags(metadata.usage_key))
-                        docs.append(doc)
-                    except Exception as err:  # pylint: disable=broad-except
-                        status_cb(f"Error indexing library component {component}: {err}")
-                    finally:
-                        num_blocks_done += 1
-                if docs:
+            docs = []
+            for component in lib_api.get_library_components(lib_key):
+                try:
+                    metadata = lib_api.LibraryXBlockMetadata.from_component(lib_key, component)
+                    doc = {}
+                    doc.update(searchable_doc_for_library_block(metadata))
+                    doc.update(searchable_doc_tags(metadata.usage_key))
+                    docs.append(doc)
+                except Exception as err:  # pylint: disable=broad-except
+                    status_cb(f"Error indexing library component {component}: {err}")
+                finally:
+                    num_blocks_done += 1
+            if docs:
+                try:
                     # Add all the docs in this library at once (usually faster than adding one at a time):
                     _wait_for_meili_task(client.index(temp_index_name).add_documents(docs))
-            except Exception as err:  # pylint: disable=broad-except
-                status_cb(f"Error indexing library {lib_key}: {err}")
-            finally:
-                num_contexts_done += 1
+                except (TypeError, KeyError, MeilisearchError) as err:
+                    status_cb(f"Error indexing library {lib_key}: {err}")
+                finally:
+                    num_contexts_done += 1

         ############## Courses ##############
         status_cb("Indexing courses...")
```

Reply:
Great solution, @pomegranited !
Fixed here: 540fbf2f30a8c7d53c9ad94b3157161f77244b9a

The `get_library_components` is outside the try/catch, but I think this is doable.